File: practice_problems.py
from typing import List, Dict, Tuple

def pig_latin(message: str):
    msg = ''
    values = message.split(' ')
    for value in values:
        if '-' in value:
            vals = value.split('-')
            val = vals[1].split('ay')[0]+vals[0]
        elif 'yay' in value and '-' not in value:
            val = value.split('yay')[0]
        msg += val + ' '
    print(msg.rstrip(' '))


def pickingNumbers(a):
    count = 0
    sets = list()
    
    def checkNumber(i, nums):
        for num in nums:
            if abs(num-i) > 1:
                return False

        return True

    while count < len(a):
        l = list()
        num = a.pop(0)
        l.append(num)
        for i in a:
            if abs(num-i) <= 1:
                if checkNumber(i, l):
                    l.append(i)
        
        a.append(num)
        sets.append(l)
        count += 1

    return max([len(s) for s in sets])

def acmTeam(topic):
    count = 0
    topic_length = len(topic)
    max_value = 0
    team_count=0
    while count < topic_length:
        t = topic.pop(0)

        if len(topic) == 0:
            break
        for item in topic:
            value = 0
            for i in range(t):
                if (int(t[i]) | int(item[i])) > 0:
                    value += 1
            if max_value < value:
                max_value = value
                team_count = 1
            elif max_value == value:
                team_count += 1
                
        count += 1
    
    return max_value, team_count

def formingMagicSquare(s: List[List[int]]):
    coords: Dict[Tuple[Tuple[int]]] = {
        0: ((0,0),(0,1),(0,2)),
        1: ((1,0),(1,1),(1,2)),
        2: ((2,0),(2,1),(2,2)),
        3: ((0,0),(1,0),(2,0)),
        4: ((0,1),(1,1),(2,1)),
        5: ((0,2),(1,2),(2,2)),
        6: ((0,0),(1,1),(2,2)),
        7: ((0,2),(1,1),(2,0))
    }

    effect_multipler: Dict[Tuple[Tuple[int]]] = {
        2: ((0,1),(1,0),(1,2),(2,1)),
        3: ((0,0),(0,2),(2,0),(2,2)),
        4: ((1,1))
    }

    def get_coord_totals(coordinates: dict, value: int=0, coord: Tuple[int] = None) -> list:
        totals = list()
        for v in coordinates.values():
            total = 0
            for i in range(len(v)):
                prev = 0
                for j, val in enumerate(v[i]):
                    if j == 0:
                        prev = val
                        continue
                    total += s[prev][val]

            totals.append(total)
            total = 0
        return totals
        

    coord_totals: list = get_coord_totals(coordinates=coords)
    sum_total: int = sum(coord_totals)
    diff: int = len(coords)-(sum_total%len(coords))

    next_magic = int((sum_total+diff)/8)
    diffs: list = [next_magic-coord for coord in coord_totals]
    
    while True:
        for i, d in enumerate(diffs):
            get_coord_totals(coordinates=coords, value=d, coord=coords[i][0])
    
    # print(sum_total%len(coords))
    return coord_totals

if __name__ == "__main__":
    # pig_latin("idden-hay essage-may")
    # pickingNumbers([1,2,2,3,1,2])
    # pickingNumbers([4,6,5,3,3,1])
    # pickingNumbers([1,1,2,2,4,4,5,5,5])
    # pickingNumbers([14,18,17,10,9,20,4,13,19,19,8,15,15,17,6,5,15,12,18,2,18,7,20,8,2,8,11,2,16,2,12,9,3,6,9,9,13,7,4,6,19,7,2,4,3,4,14,3,4,9,17,9,4,20,10,16,12,1,16,4,15,15,9,13,6,3,8,4,7,14,16,18,20,11,20,14,20,12,15,4,5,10,10,20,11,18,5,20,13,4,18,1,14,3,20,19,14,2,5,13])

    formingMagicSquare(s=[[4,8,2],[4,5,7],[6,1,6]])
    # formingMagicSquare(s=[[4,9,2],[3,5,7],[8,1,6]])
    formingMagicSquare(s= [[5, 3, 4], [1, 5, 8], [6, 4, 2]])
    # formingMagicSquare(s= [[8, 3, 4], [1, 5, 9], [6, 7, 2]])
    formingMagicSquare(s= [[4,9,2],[3,5,7],[8,1,5]])
    # formingMagicSquare(s= [[4,9,2],[3,5,7],[8,1,6]])
    

File: practiceProblems.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func addItUp(number int) int {
	sum := 0
	for i := 1; i <= number; i++ {
		sum += i
	}
	return sum
}

func ceasarCipher(msg string, shift int) string {
	alphabet := map[rune]int{
		'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5, 'f': 6, 'g': 7, 'h': 8, 'i': 9, 'j': 10, 'k': 11, 'l': 12, 'm': 13, 'n': 14, 'o': 15, 'p': 16,
		'q': 17, 'r': 18, 's': 19, 't': 20, 'u': 21, 'v': 22, 'w': 23, 'x': 24, 'y': 25, 'z': 26,
	}
	runeMsg := []rune(msg)
	resultStr := ""
	newIndex := 0
	for _, char := range runeMsg {
		if index, ok := alphabet[char]; ok {
			if index+shift > 26 {
				newIndex = shift - (26 - index)
			} else {
				newIndex = index + shift
			}
			for key, value := range alphabet {
				if value == newIndex {
					resultStr += string(key)
				}
			}
			continue
		}
		resultStr += string(char)
	}
	return resultStr
}

func acmTeam(topic []string) []int32 {
	maxValue := 0
	teamCount := 0
	for _, list := range topic {
		for _, item := range topic[1:] {
			value := 0
			for i := range []int32(list) {
				if (int32(list[i]) | int32(item[i])) == 49 {
					value++
				}
			}
			if maxValue < value {
				maxValue = value
				teamCount = 1
			} else if maxValue == value {
				teamCount++
			}
		}
		topic = topic[1:]
	}
	return []int32{int32(maxValue), int32(teamCount)}
}

func nonDivisibleSubset(k int32, s []int32) int32 {
	// Write your code here
	sets := []map[int32]struct{}{}

	var checkNum func(int32, map[int32]struct{}) bool = func(num int32, numbers map[int32]struct{}) bool {
		for key := range numbers {
			if (num+key)%k == 0 {
				return false
			}
		}
		return true
	}

	var max int32 = 0
	for _, num := range s {
		nonDivisible := make(map[int32]struct{})
		s = s[1:]
		nonDivisible[num] = struct{}{}

		for _, item := range s {
			if (num+item)%k != 0 {
				if checkNum(item, nonDivisible) {
					nonDivisible[item] = struct{}{}
				}
			}
		}
		s = append(s, num)

		length := int32(len(nonDivisible))
		if length > max {
			max = length
		}
		sets = append(sets, nonDivisible)
	}

	return max
}

func taumBday(b int32, w int32, bc int32, wc int32, z int32) int64 {
	// Write your code here
	var total int64 = 0
	if bc > (z + wc) {
		total = int64(b) * int64((wc + z))
		total += int64(wc) * int64(w)
	} else if wc > (z + bc) {
		total = int64(w) * int64(bc+z)
		total += int64(bc) * int64(b)
	} else {
		total = int64(b)*int64(bc) + int64(w)*int64(wc)
	}

	return total
}

// Complete the countSort function below.
func countSort(arr [][]string) {
	mid := len(arr) / 2

	output := ""
	for i := 1; i < len(arr); i++ {
		key := arr[i]
		j := i - 1

		if j <= mid {
			arr[j][1] = "-"
		}

		for j > 0 && key[0] < arr[j][0] {
			arr[j+1] = arr[j]
			j--
		}

		if i < mid {
			key[1] = "-"
		}

		// if j < 0 {
		// 	fmt.Println(j)
		// }

		if arr[j][0] == key[0] {
			arr[j] = append(arr[j], key[1])
			left := arr[:j+1]
			right := arr[j+2:]

			arr = left
			arr = append(arr, right...)
			i--
			mid--
		} else {
			arr[j+1] = key
		}
	}

	for i := range arr {
		for j := 1; j < len(arr[i]); j++ {
			output = fmt.Sprintf("%s%s ", output, arr[i][j])
		}
	}
	fmt.Println(output)
}

func readFileIntoMemory(name string) [][]string {
	results := [][]string{}

	text, _ := os.Open(name)
	defer text.Close()

	reader := bufio.NewScanner(text)

	i := 0
	for reader.Scan() {
		line := reader.Text()

		results = append(results, []string{})
		for _, split := range strings.Split(line, " ") {
			results[i] = append(results[i], split)
		}
		i++
	}
	return results
}

func main() {
	// fmt.Println(addItUp(10))
	// fmt.Println(ceasarCipher("abcd xyz", 4))
	// fmt.Println(acmTeam([]string{`10101`, "11110", "00010"}))

	// fmt.Println(nonDivisibleSubset(4, []int32{19, 10, 12, 24, 25, 22}))

	// fmt.Println(taumBday(27984, 1402, 619246, 615589, 247954))

	// countSort([][]string{{"0", "e"}, {"2", "a"}, {"1", "b"}, {"3", "a"}, {"4", "f"}, {"1", "f"}, {"2", "a"}, {"1", "e"}, {"1", "b"}, {"1", "c"}})
	// countSort([][]string{{"0", "ab"}, {"6", "cd"}, {"0", "ef"}, {"6", "gh"}, {"4", "ij"}, {"0", "ab"}, {"6", "cd"}, {"0", "ef"}, {"6", "gh"}, {"0", "ij"},
	// 	{"4", "that"}, {"3", "be"}, {"0", "to"}, {"1", "be"}, {"5", "question"}, {"1", "or"}, {"2", "not"}, {"4", "is"}, {"2", "to"}, {"4", "the"}})
	countSort(readFileIntoMemory("countingsort.txt"))
}
